### PR TITLE
debuginfo: embed gdb visualizers in module

### LINKED ISSUE #1386: Implement #![debugger_visualizer]
https://doc.rust-lang.org/nightly/reference/attributes/debugger.html#using-debugger_visualizer-with-gdb
https://github.com/rust-lang/rust/blob/ab0f3e680e4fd50218914b8afa6900a67377721c/compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs

The `.debug_gdb_scripts` section should contain `\x01gdb_load_rust_pretty_printers.py` followed by for each `#![debugger_visualizer(gdb_script_file = "...")]` the pretty printer inlined into the executable. Except if one of the conditions to omit `.debug_gdb_scripts` is true.

### PR DESCRIPTION
**Pull request by kartva**
fixes #1386

### CODE REVIEW

File: src/debuginfo/emit.rs [hunk 1]
```diff
@@ -35,6 +35,23 @@ impl DebugContext {
             }
             Ok(())
         });
+
+        self.add_gdb_debugger_visualizers(product);
```
Comment (bjorn3) [inline src/debuginfo/emit.rs:39]: This function needs to be called at https://github.com/rust-lang/rustc_codegen_cranelift/blob/8c46d93ca0f287ea3b284c061eebcb4a275fc43a/src/main_shim.rs#L40 and a load of the produced global needs to be inserted there too to ensure the section only gets inserted a single time in the whole program and that it doesn't get removed by the linker.

File: src/debuginfo/emit.rs [hunk 2]
```diff
@@ -35,6 +35,23 @@ impl DebugContext {
...
+    }
+
+    fn add_gdb_debugger_visualizers(&mut self, product: &mut ObjectProduct) {
```
Comment (bjorn3) [inline src/debuginfo/emit.rs:42]: If https://github.com/rust-lang/rust/blob/7606c13961ddc1174b70638e934df0439b7dc515/compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs#L85-L117 returns false you need to skip adding the section.

File: src/debuginfo/emit.rs [hunk 3]
```diff
@@ -35,6 +35,23 @@ impl DebugContext {
...
+        let mut section = WriterRelocate::new(self.endian);
```
Comment (bjorn3) [inline src/debuginfo/emit.rs:43]: No need to go through `WriterRelocate` here. A plain `Vec` is enough. There will be no relocations inside this section.

**Comment by ChiefMateStarbuck** [comment]
Is this PR still needed? I can pick this up and clean it up.

**Comment by bjorn3** [comment]
Yes, feel free to pick it up.

**Comment by kernel-loophole** [comment]
@ChiefMateStarbuck are you working on it?

**Comment by ChiefMateStarbuck** [comment]
@kernel-loophole , hello. I am not currently working on this. You are free to pick it up.

**Comment by kernel-loophole** [comment]
thanks @ChiefMateStarbuck

**Pull request closed without merging**